day_10/task_2/pipe_maze.cpp

```diff
@@ -270,3 +276,86 @@ void AttributedMaze::color_neighbor(const MazePoint& loop_point, const MazePoint
         }
     }
 }
+
+void AttributedMaze::print_states()
+{
+    for(const auto& row : maze_)
+    {
+        for(const auto& single_tile : row)
+        {
+            if(single_tile.state == State::Undefined)
+            {
+                std::cout << ".";
+            }
+            if(single_tile.state == State::LeftColor)
+            {
+                std::cout << "L";
+            }
+            if(single_tile.state == State::RightColor)
+            {
+                std::cout << "R";
+            }
+            if(single_tile.state == State::Loop)
+            {
+                std::cout << "X";
+            }
+
+        }
+        std::cout << std::endl;
+    }
+}
+
+void AttributedMaze::color_neighbors()
+{
+    const MazePoint starting_point{find_starting_point()};
+    MazePoint prev, current;
+    prev = starting_point;
+
+    current = find_next_neighbors(starting_point).first;
+    color_neighbor(prev, current);
+
+    while(current != starting_point)
+    {
+        auto temp = go_forward(current, prev);
+        prev = current;
+        current = temp;
+        color_neighbor(prev, current);
+    }
+}
+
+State AttributedMaze::find_which_color_is_closed()
+{
+    for(auto i = 0U; i < maze_.size(); i++)
+    {
+        for(auto j = 0U; j < maze_[i].size(); j++)
+        {
+            if(maze_[i][j].state == State::LeftColor)
+            {
+                return State::RightColor;
+            }
+            if(maze_[i][j].state == State::RightColor)
+            {
+                return State::LeftColor;
+            }
+        }
+    }
+    return State::Undefined;
+}
```

Comment:
```suggestion
State AttributedMaze::find_which_color_is_closed()
{
    for(auto i = 0U; i < maze_.size(); i++)
    {
        for(auto j = 0U; j < maze_[i].size(); j++)
        {
            if(maze_[i][j].state == State::LeftColor)
            {
                return State::RightColor;
            }
            if(maze_[i][j].state == State::RightColor)
            {
                return State::LeftColor;
            }
        }
    }
    return State::Undefined;
}
```

[optional]

Please consider the version of algorithm which would cover all possible input cases